Added linalg.eigh, linalg.eigvalsh

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.eigh`, and `torch.linalg.eigvalsh` for NumPy compatibility.
The current `torch.symeig` uses (on CPU) a different LAPACK routine than NumPy (`syev` vs `syevd`). Even though it shouldn't matter in practice, `torch.linalg.eigh` uses `syevd` (as NumPy does).

Ref #42666

IvanYashchuk committed:
Started torch.linalg.eigh and torch.linalg.eigvalsh

Added wrapper for CPU syevd.

IvanYashchuk committed:
Fixed syevd cpu

IvanYashchuk committed:
Added test for linalg.eigh

IvanYashchuk committed:
Expose linalg_eigh, linalg_eigvalsh functions in torch.linalg

IvanYashchuk committed:
Use smaller shape

IvanYashchuk committed:
Added linalg_eigh documentation

IvanYashchuk committed:
Added linalg_eigvalsh documentation

IvanYashchuk committed:
Remove code duplication due to rebasing

IvanYashchuk committed:
Updated test_eigh

IvanYashchuk committed:
Added test_eigvalsh

IvanYashchuk committed:
Remove c10::optional uplo

IvanYashchuk committed:
Added _syevd_helper_cuda

Upadted tests

IvanYashchuk committed:
Use (void) to mark unused arguments.

Updated apply_syevd

nikitaved commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8224,6 +8224,23 @@
   use_c10_dispatcher: full
   variants: function, method
 
+- func: _syevd_helper(Tensor self, bool compute_v, str uplo) -> (Tensor, Tensor)
```
Could you please replace the parameter names with something more common, such as `compute_v -> compute_eigenvectors`, `uplo -> upper`.

nikitaved commented on test/test_linalg.py:
```diff
@@ -180,6 +180,54 @@ def test_det(self, device, dtype):
+            expected_w, expected_v = np.linalg.eigh(matrix.cpu().numpy())
+            actual_w, actual_v = torch.linalg.eigh(matrix)
+            self.assertEqual(actual_w, expected_w)
+            # sign of eigenvectors is not unique and therefore absolute values are compared
```
You can make signs deterministic by multiplying each eigenvector with a sing of the largest in absolute real or complex value coordinate, similar to how it is done in `test_lobpcg` in `test_autograd.py`.

IvanYashchuk replied:
I could write something like
```python
_, expected_v = np.linalg.eigh(matrix.cpu().numpy())
_, actual_v = torch.linalg.eigh(matrix)
# get the sign for each column
_, idx = actual_v.abs().max(-2, keepdim=True)
sign = actual_v.gather(-2, idx).sign()
is_equal = np.allclose((sign * actual_v).cpu().numpy(), expected_v) or np.allclose(actual_v.cpu().numpy(), expected_v)
self.assertTrue(is_equal)
```
This would require an additional if for a matrix with `(0, 0)` shape, and some tweaking since `.sign()` does not work for complex numbers.
Effectively I think the above code doesn't do better in comparing the numbers to NumPy output than
```python
self.assertTrue(abs(actual_v), abs(expected_v))
```

nikitaved replied:
`sing` could be run on either real or complex part. Up to you, just saying, but `abs(actual_v)` will return `|z|`, not `|z| e^{i * \phi}` for each entry, so we are not testing the angular information.

IvanYashchuk replied:
I've added the test that compares eigenvectors with the flipped sign, thank you for the suggestion!

nikitaved commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +142,93 @@
     (tensor(3.7417), tensor(11.2250))
 """)
+
+_add_docstr(_linalg.linalg_eigh, r"""
+linalg.eigh(input, UPLO='L') -> tuple(Tensor, Tensor)
```
`UPLO -> upper`.

IvanYashchuk replied:
`UPLO={'L', 'U'}` is used for for NumPy compatibility.

nikitaved commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +142,93 @@
+This function calculates all eigenvalues (and vectors) of :attr:`input`
+such that :math:`\text{input} = V \text{diag}(e) V^H`.
+
+Since the input matrix :attr:`input` is supposed to be Hermitian,
+only the lower triangular portion is used by default
```
`symeig` assumes the upper part by default, could you please make it consistent with `symeig`?

IvanYashchuk replied:
The aim is to be compatible with NumPy, `numpy.linalg.eigh` considers the lower part by default.

IvanYashchuk committed:
Renamed compute_v -> compute_eigenvectors

IvanYashchuk committed:
Added test check with sign flipping of eigenvectors

nikitaved reviewed (approved):
Looks good to me. My issues were addressed.

mruberry commented:
@IvanYashchuk Is this ready for review?

IvanYashchuk commented:
Not yet, I think it requires slightly a bit more work now to bring it to the same shape as other PRs to `torch.linalg`

mruberry commented:
> Not yet, I think it requires slightly a bit more work now to bring it to the same shape as other PRs to `torch.linalg`

Sounds good. Let me know.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-eig

IvanYashchuk committed:
Remove skip if numpy not available

IvanYashchuk committed:
Update input argument description

IvanYashchuk committed:
Added non-contiguous inptu tests

IvanYashchuk committed:
Updated docs

IvanYashchuk committed:
Make eigh return named tensors, add default value for uplo

IvanYashchuk committed:
Updated UPLO argument implementation

IvanYashchuk committed:
Added out= variant

IvanYashchuk committed:
Added wrappers to torch/linalg.h

IvanYashchuk committed:
Added overrides.py entry

IvanYashchuk committed:
Added derivative rules for linalg_eigh

IvanYashchuk committed:
Check dtypes for out= variants

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-eig

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -183,6 +183,266 @@ def test_det(self, device, dtype):
+    @skipCPUIfNoLapack
+    @dtypes(torch.float64, torch.complex128)
+    @dtypesIfCUDA(torch.float64)
+    def test_eigh_autograd(self, device, dtype):
```
Testing of `eigh` differentiation requires a modified loss function, `torch.linalg.eigh` is not tested directly. I think it can't be implemented with `method_tests`.

mruberry replied:
That's OK. We'll soon switch to using OpInfo-based tests and can revisit this.

IvanYashchuk commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8876,6 +8876,37 @@
+  dispatch:
+    DefaultBackend: linalg_eigvalsh
+
+- func: linalg_eigvalsh.out(Tensor self, str UPLO='L', *, Tensor(a!) out) -> Tensor(a!)
```
`'L'` is used here instead of `"L"` as it caused errors during the compilation. I didn't investigate it further, I hope it's fine and doesn't break anything.

IvanYashchuk committed:
flake8

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-eig

IvanYashchuk committed:
Add namedtuple entry

IvanYashchuk committed:
Fix typo

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-eig

IvanYashchuk committed:
Added entry for linalg_eigh namedtuple test

IvanYashchuk committed:
Merge branch 'master' into numpy-eig

IvanYashchuk commented:
@mruberry, this PR is now ready for review.

The current `torch.symeig` uses (on CPU) a different LAPACK routine than NumPy (`syev` vs `syevd`). This PR adds `syevd` LAPACK wrappers for CPU `linalg_eigh` and `linalg_eigvalsh`. For GPU, the same backend function as for `symeig` is used.

`torch.linalg.eigh` returns now a named tuple, as `torch.symeig` does. Is that what we want, or should it be just a tuple?
So that tests for named tuple are not modified much, I added tensor method for `linalg_eigh` (and for completeness for `linalg_eigvalsh`). Is that okay?

Differentiation of `linalg.eigh` is supported and tested with a separate test `test_linalg.py`. For finite differences check in `gradcheck` to work correctly we need to test a modified input function not `linalg.eigh` directly, that is not supported with `method_tests` in `test_autograd.py`.
`linalg.eigvalsh` is marked as non-differentiable, since the implementation requires computed eigenvectors. It's noted in the documentation and suggested to use `linalg.eigh` instead.

mruberry commented:
> @mruberry, this PR is now ready for review.
> 
> The current `torch.symeig` uses (on CPU) a different LAPACK routine than NumPy (`syev` vs `syevd`). This PR adds `syevd` LAPACK wrappers for CPU `linalg_eigh` and `linalg_eigvalsh`. For GPU, the same backend function as for `symeig` is used.

Sounds great!
>
> `torch.linalg.eigh` returns now a named tuple, as `torch.symeig` does. Is that what we want, or should it be just a tuple?

A named tuple seems strictly better.

> So that tests for named tuple are not modified much, I added tensor method for `linalg_eigh` (and for completeness for `linalg_eigvalsh`). Is that okay?

Let's hold off on adding method variants for the moment since other linalg functions don't have them. We'll revisit this in the future once we've cleaned up the existing linalg functions and their method variants.

> Differentiation of `linalg.eigh` is supported and tested with a separate test `test_linalg.py`. For finite differences check in `gradcheck` to work correctly we need to test a modified input function not `linalg.eigh` directly, that is not supported with `method_tests` in `test_autograd.py`.

OK. I'll take a closer look but that sounds sensible. 

> `linalg.eigvalsh` is marked as non-differentiable, since the implementation requires computed eigenvectors. It's noted in the documentation and suggested to use `linalg.eigh` instead.

Sounds like a good start.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+  char uplo = uplo_str == "U" ? 'U' : 'L';
+  char jobz = compute_v ? 'V' : 'N';
+
+  int info;
```
Add a comment here explaining that the use of imprecise types, like int, is consistent with the LAPACK interface (otherwise we'd use a precise type, like int32_t).

mruberry replied:
Nit: if we were being extremely correct we would add a using declaration to make it easy to change this int type later

IvanYashchuk replied:
I think this kind of type change should happen at once for all functions that call to LAPACK. There should be `lapack_int` that depends on the type of the linked LAPACK or only 64-bit indexing should be used by default and only 64-bit version of LAPACK allowed.

mruberry replied:
I agree with you and the comment looks great.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+  lapackSyevd<scalar_t, value_t>(jobz, uplo, n, v_data, lda, w_data, &work_query, lwork, &rwork_query, lrwork, &iwork_query, liwork, &info);
+  lwork = std::max(1, static_cast<int>(real_impl<scalar_t, value_t>(work_query)));
+  Tensor work = at::empty({lwork}, v.options());
+  liwork = std::max(1, static_cast<int>(iwork_query));
```
The static cast here is redundant with the declaration of iwork_query above. I think you want `std::max<int>` instead?

IvanYashchuk replied:
True. I've changed that.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+  int iwork_query;
+
+  lapackSyevd<scalar_t, value_t>(jobz, uplo, n, v_data, lda, w_data, &work_query, lwork, &rwork_query, lrwork, &iwork_query, liwork, &info);
+  lwork = std::max(1, static_cast<int>(real_impl<scalar_t, value_t>(work_query)));
```
As below, probably want to be explicitly about the type for std::max

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+  Tensor rwork;
+  value_t* rwork_data = nullptr;
+  if (isComplexType(at::typeMetaToScalarType(v.dtype()))) {
+    lrwork = std::max(1, static_cast<int>(rwork_query));
```
Another note on specifying the type of std::max explicitly

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+    rwork_data = rwork.data_ptr<value_t>();
+  }
+
+  for (int64_t i = 0; i < batch_size; i++) {
```
Add a comment here explaining what this for loop does

mruberry replied:
The index variable `i` should be typed like `batch_size`:

```
for (auto i = decltype(batch_size){0}; i < batch_size; i++) {
```

IvanYashchuk replied:
Added a comment that this loop calls `lapackSyevd` for each matrix in the batched input. `lapackSyevd` works only for a single matrix.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+    lapackSyevd<scalar_t, value_t>(jobz, uplo, n, v_working_ptr, lda, w_working_ptr, work.data_ptr<scalar_t>(), lwork, rwork_data, lrwork, iwork.data_ptr<int>(), liwork, &info);
+    infos[i] = info;
+    if (info != 0) {
+      return;
```
Add a comment here explaining what triggers this return, and what callers should do if this is triggered

mruberry replied:
Follow-up question:

In the future we expect to add `_ex` variants of functions in linear algebra that have additional error information. For example, let's say we have `torch.linalg.foo` that computes the `foo` operation on a non-singular matrix. If the matrix is singular, however, then the math library we're using returns error information on a device tensor explaining that the matrix is singular. 

On the CPU, torch.linalg.foo(cpu_matrix) can throw an error when a singular tensor is encountered. On CUDA, however, translating the device tensor containing the error information requires a cross-device sync, which we try to avoid. torch.linalg.foo(cuda_matrix) should perform and document this sync, however.

torch.linalg.foo_ex, however, will return both device tensors and not perform the error check. This allows the user to check the error themselves, and avoid an immediate cross-device sync on CUDA if they want.

IvanYashchuk replied:
The current behavior for linalg (LAPACK/MAGMA based) functions to raise an error if something goes wrong or input doesn't satisfy some requirement therefore all batched CPU functions return early since further computations will be wasted anyway. In the future, this early return should be removed.

What does `_ex` suffix mean?
I agree that this kind of change will improve performance. I think the default behavior should be the same on CPU and CUDA, so CPU shouldn't raise an error, even though it's cheap to check for it, if CUDA doesn't.
What do you think about instead of introducing new functions with `_ex` suffix in Python interface modify existing ones to accept optional `info=` argument, similar to `out=`, that will be filled with the returned LAPACK/MAGMA/cuSOLVER error codes for the user to check if he wants to.
```python
eigenvalues, eigenvectors = torch.linalg.eigh(input, info=torch.empty(batch_size))
# instead of 
eigenvalues, eigenvectors, info = torch.linalg.eigh_ex(input)
```

mruberry replied:
> The current behavior for linalg (LAPACK/MAGMA based) functions to raise an error if something goes wrong or input doesn't satisfy some requirement therefore all batched CPU functions return early since further computations will be wasted anyway. In the future, this early return should be removed.

Sounds good for now.

 
> What does `_ex` suffix mean?

"Extra."

> I agree that this kind of change will improve performance. I think the default behavior should be the same on CPU and CUDA, so CPU shouldn't raise an error, even though it's cheap to check for it, if CUDA doesn't.

Agreed. Both the CPU and CUDA ex variants will not perform their own error checking and return the info tensor for callers to evaluate.

> What do you think about instead of introducing new functions with `_ex` suffix in Python interface modify existing ones to accept optional `info=` argument, similar to `out=`, that will be filled with the returned LAPACK/MAGMA/cuSOLVER error codes for the user to check if he wants to.

This is a really good idea and we did consider an approach like this, but there are technical issues with this approach that make it prohibitive.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ syevd ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_syevd(Tensor& w, Tensor& v, bool compute_v, std::string uplo_str, std::vector<int64_t>& infos) {
```
`std::string uplo_str` -> `const std::string& uplo_str` ?

IvanYashchuk replied:
Changed it here, but for other places it can't be done because the signature of the autogenerated code is `std::string` not `const std::string&`.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+#endif
+}
+
+std::tuple<Tensor, Tensor> _syevd_helper_cpu(const Tensor& self, bool compute_eigenvectors, std::string uplo) {
```
`const std::string&` here, too

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+#endif
+}
+
+std::tuple<Tensor, Tensor> _syevd_helper_cpu(const Tensor& self, bool compute_eigenvectors, std::string uplo) {
```
Add a comment explaining what this function does and how it's intended to be called

IvanYashchuk replied:
Added:
```cpp
// This function computes eigenvalues 'w' and eigenvectors 'v' of the tensor 'self'
// compute_eigenvectors controls whether eigenvectors should be computed
// uplo controls the portion of input matrix to consider in computations, allowed values are "u", "U", "l", "L"
// This function prepares correct input for 'apply_syevd' and checks for possible errors using 'infos'
```

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ syevd ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_syevd(Tensor& w, Tensor& v, bool compute_v, std::string uplo_str, std::vector<int64_t>& infos) {
```
Add a comment describing the function and how it's intended to be called

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+  return at::_syevd_helper(self, /*compute_eigenvectors=*/true, uplo);
+}
+
+std::tuple<Tensor&, Tensor&> linalg_eigh_out(Tensor& eigvals, Tensor& eigvecs, const Tensor& self, std::string uplo) {
```
It looks like it would be possible to avoid the copy and use the already allocated storage space directly. This PR doesn't have to do that, but should add a TODO comment

IvanYashchuk replied:
Yes, that's right it should be fairly easy to do that. I'd do it in a follow-up PR.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -863,6 +893,134 @@ std::tuple<Tensor&,Tensor&> qr_out(Tensor& Q, Tensor& R, const Tensor& self, boo
+  return eigvals;
+}
+
+Tensor& linalg_eigvalsh_out(Tensor& result, const Tensor& self, std::string uplo) {
```
Same TODO to avoid the copy here, I think

mruberry commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -333,4 +333,10 @@ static inline int64_t computeLRWorkDim(const char jobz, int64_t m, int64_t n) {
   return std::max(5 * mn * mn + 5 * mn, 2 * mx * mn + 2 * mn * mn + mn);
 }
 
+static inline void checkUplo(std::string uplo) {
```
Add a comment describing this function and its intended use. `const std::string&`?

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1690,6 +1690,13 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ syevd ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+std::tuple<Tensor, Tensor> _syevd_helper_cuda(const Tensor& self, bool compute_eigenvectors, std::string uplo) {
```
`const std::string&`. Add a comment explaining this function and how it's intended to be used.

IvanYashchuk replied:
`const std::string&` can be used, since the autogenerated function uses `std::string`
```
// aten::_syevd_helper(Tensor self, bool compute_eigenvectors, str uplo) -> (Tensor, Tensor)
std::tuple<Tensor,Tensor> _syevd_helper(const Tensor & self, bool compute_eigenvectors, std::string uplo) {
    static auto op = c10::Dispatcher::singleton()
        .findSchemaOrThrow("aten::_syevd_helper", "")
        .typed<std::tuple<Tensor,Tensor> (const Tensor &, bool, std::string)>();
    return op.call(self, compute_eigenvectors, uplo);
}
```

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8918,6 +8918,37 @@
+- func: linalg_eigvalsh(Tensor self, str UPLO="L") -> Tensor
+  python_module: linalg
+  use_c10_dispatcher: full
+  variants: method, function
```
Per discussion, let's remove these method variants for now

mruberry commented on test/test_linalg.py:
```diff
@@ -185,6 +185,264 @@ def test_det(self, device, dtype):
+            expected_w, expected_v = np.linalg.eigh(matrix.cpu().numpy(), UPLO=uplo)
+            actual_w, actual_v = torch.linalg.eigh(matrix, UPLO=uplo)
+            self.assertEqual(actual_w, expected_w)
+            # sign of eigenvectors is not unique and therefore absolute values are compared
```
Where are we divergent? Only on CUDA?

IvanYashchuk replied:
CPU as well, discrepancies could be due to NumPy computing results in double precision or linked to a different LAPACK library.

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -1983,6 +1983,20 @@ def add_docstr_all(method, docstr):
 See :func:`torch.kthvalue`
 """)
 
+add_docstr_all('linalg_eigh',
```
Once the method variants are removed these would be, too.

Note that for these to render properly they would have needed entries in tensors.rst, too.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+eigh = _add_docstr(_linalg.linalg_eigh, r"""
+linalg.eigh(input, UPLO='L') -> tuple(Tensor, Tensor)
+
+This function returns eigenvalues and eigenvectors
```
Does this function work on batches of these tensors, too?

mruberry replied:
"This function returns" -> "Computes the"

IvanYashchuk replied:
It works for batched input as well, that's tested and I updated the docs now.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+represented by a namedtuple (eigenvalues, eigenvectors).
+
+This function calculates all eigenvalues (and vectors) of :attr:`input`
+such that :math:`\text{input} = V \text{diag}(e) V^H`, where :math:`e` is eigenvalues tensor
```
"eigenvalues tensor" -> "tensor of eigenvalues"
"eigenvectors tensor" -> "tensor of eigenvectors"

mruberry replied:
This second paragraph is mostly redundant with the first. Can they be combined?

mruberry replied:
UPLO should be mentioned here and in the mathematics, too.

IvanYashchuk replied:
Changed now to
```
This function computes the eigenvalues and eigenvectors
of a complex Hermitian (or real symmetric) matrix :attr:`input` or batches of such matrices
such that :math:`\text{input} = V \text{diag}(w) V^H`, where :math:`w` is the tensor of eigenvalues,
:math:`V` is the tensor of eigenvectors and :math:`^H` is the conjugate transpose operation.

Since :attr:`input` is assumed to be consisting of Hermitian matrices,
only the lower triangular portion of these matrices is used by default (``UPLO = 'L'``)  in computations
and the imaginary part of the diagonal will always be treated as zero.
```

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+such that :math:`\text{input} = V \text{diag}(e) V^H`, where :math:`e` is eigenvalues tensor
+and :math:`V` is eigenvectors tensor.
+
+Since the input matrix :attr:`input` is supposed to be Hermitian,
```
"is supposed" -> "is assumed"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+.. note:: The eigenvalues of real symmetric or complex Hermitian matrices are always real.
+
+.. note:: The eigenvalues/eigenvectors are computed using LAPACK/MAGMA routines ``_syevd``, ``_heevd``.
+
```
This function causes a cross-device sync on CUDA, right? If so, this PR should add a warning here explaining that.

mruberry replied:
That warning can be combined with a warning that this function will throw an error if a tensor does not meet the requirements.

IvanYashchuk replied:
I added a note on cross-device synchronization for CUDA.

This function doesn't throw errors if the input is not Hermitian or symmetric because of

> Since :attr:`input` is assumed to be consisting of Hermitian matrices,
only the lower triangular portion of these matrices is used by default (``UPLO = 'L'``)  in computations
and the imaginary part of the diagonal will always be treated as zero.

[The documentation says](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem-routines/lapack-least-squares-and-eigenvalue-problem-driver-routines/symmetric-eigenvalue-problems-lapack-driver-routines/syevd.html) the error code can be >0 indicating that the algorithm didn't converge. But I don't know what kind of input is needed to make it fail.

mruberry replied:
Interesting... is our failure behavior consistent with NumPy's, then?

IvanYashchuk replied:
Yes. NumPy raises `LinAlgError("Eigenvalues did not converge")` for `info > 0`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+.. note:: The eigenvalues/eigenvectors are computed using LAPACK/MAGMA routines ``_syevd``, ``_heevd``.
+
+Args:
+    input (Tensor): the input tensor of size :math:`(*, n, n)` consisting of Hermitian
```
In another PR the cases for a square matrix (n, n) and a batch of such matrices were separated. That pattern might be helpful to follow here.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+Args:
+    input (Tensor): the input tensor of size :math:`(*, n, n)` consisting of Hermitian
+                    :math:`n \times n` matrices, where `*` is zero or more batch dimensions.
+    UPLO ('L', 'U', optional): controls whether to consider upper-triangular or lower-triangular part.
```
The UPLO argument needs to be mentioned in the description of the function, too. This should be more specific, too. "considers" is too vague a term.

IvanYashchuk replied:
So in the main text, it says now "only the lower triangular portion of these matrices is used by default (`UPLO = 'L'`) in computations".
I changed "considers" -> "controls whether to use upper-triangular or lower-triangular part of :attr:`input` in the computations."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+Since the input matrix :attr:`input` is supposed to be Hermitian,
+only the lower triangular portion is used by default
+and the imaginary part of the diagonal will always be treated as zero.
+
```
Another note expressing that the signs are not unique would be helpful. This also needs a "Supports..." statement that clarifies that CUDA complex backward is not supported.

mruberry replied:
Also there should be a pointer to eigvalsh for those that only want to compute the eigen values.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+    (Tensor, Tensor): A namedtuple (eigenvalues, eigenvectors) containing
+
+        - **eigenvalues** (*Tensor*): Shape :math:`(*, m)`.
+            The eigenvalues in ascending order, each repeated according to its multiplicity.
```
Is the phrase "each repeated according to its multiplicity" necessary?

IvanYashchuk replied:
I guess not

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+eigvalsh = _add_docstr(_linalg.linalg_eigvalsh, r"""
+linalg.eigvalsh(input, UPLO='L') -> Tensor
+
+This function returns eigenvalues of a complex Hermitian (conjugate symmetric)
```
"This function returns" -> "Computes the"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+linalg.eigvalsh(input, UPLO='L') -> Tensor
+
+This function returns eigenvalues of a complex Hermitian (conjugate symmetric)
+or real symmetric matrix :attr:`input`. The eigenvalues are returned in ascending order,
```
Same question about batch support as with eigh

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+
+.. note:: The eigenvalues are computed using LAPACK/MAGMA routines ``_syevd``, ``_heevd``.
+
+.. note:: This function doesn't support backpropagation, please use :func:`torch.linalg.eigh` instead.
```
Same comments about notes like cross-device datamovement

mruberry replied:
Extend this note to explain that eigvalsh also computes the eigenvectors

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,110 @@
+.. note:: This function doesn't support backpropagation, please use :func:`torch.linalg.eigh` instead.
+
+Args:
+    input (Tensor): the input tensor of size :math:`(*, n, n)` consisting of Hermitian
```
Same notes here as above.

One option for the documentation of eigvalsh is to simply refer to the documentation for eigh and explain that this is non differentiable and that it only returns the eigen values. That might makes things simpler and ensure consistency.

mruberry reviewed (commented):
Hey @IvanYashchuk!

This looks pretty good. I really appreciate this PR's thorough testing.

I made some comments and asked a few questions.

IvanYashchuk committed:
Typed std::max; added a few comments on the implementation

IvanYashchuk committed:
Remove method variant

IvanYashchuk committed:
Test torch.linalg.eigh namedtuple

IvanYashchuk committed:
Added test with lower case uplo

IvanYashchuk committed:
Added a few comments describing new functions

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-eig

IvanYashchuk committed:
Doc fixes

IvanYashchuk committed:
Remove unused import

IvanYashchuk committed:
Added a note on non-uniqueness of eigenvectors

IvanYashchuk commented:
@mruberry, I've added a few comments to the implementation code and updated the documentation according to the feedback.
I think it's ready now for another review.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+linalg.eigh(input, UPLO='L') -> tuple(Tensor, Tensor)
+
+This function computes the eigenvalues and eigenvectors
+of a complex Hermitian (or real symmetric) matrix :attr:`input` or batches of such matrices
```
":attr:\`input\` or batches of" -> ":attr:\`input\`, or a batch of"

IvanYashchuk replied:
Thanks! I only changed "eigenvectors :math:`V` is such that" -> "eigenvectors :math:`V` decompose the :attr:`input` such that"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+only the lower triangular portion of these matrices is used by default (``UPLO = 'L'``)  in computations
+and the imaginary part of the diagonal will always be treated as zero.
+
+Supports real-valued and complex-valued input.
```
"real-valued" -> "float, double"

mruberry replied:
(because real-value would also include half and long, etc.)

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+such that :math:`\text{input} = V \text{diag}(w) V^H`, where :math:`w` is the tensor of eigenvalues,
+:math:`V` is the tensor of eigenvectors and :math:`^H` is the conjugate transpose operation.
+
+Since :attr:`input` is assumed to be consisting of Hermitian matrices,
```
This paragraph is already pretty good, but we can probably rearrange it to highlight the "UPLO" argument. What about something like:

"Since the matrix or matrices in :attr:\`input\` are assumed to be Hermitian, the imaginary part of their diagonals is always treated as zero. When :attr:\`UPLO\` is "L", its default value, only the lower triangular part of each matrix is used in the computation. When :attr:\`UPLO\` is "U" only the upper triangular part of each matrix is used."

IvanYashchuk replied:
Thanks! I've used your suggestion.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+.. note:: The eigenvalues of real symmetric or complex Hermitian matrices are always real.
+
+.. note:: The eigenvectors of matrices are not unique, any eigenvector multiplied by a constant remains
+          to be a valid eigenvector, therefore results on CPU, CUDA and from other libraries can be different.
```
"not unique," -> "not unique, so"

mruberry replied:
"remains to be" -> "remains"

mruberry replied:
"eigenvector, therefore results on CPU, CUDA and from other libraries can be different." -> "eigenvector. This function may compute different eigenvector representations on different device types."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+          to be a valid eigenvector, therefore results on CPU, CUDA and from other libraries can be different.
+          Usually the difference is only in the sign of the eigenvector.
+
+.. note:: The eigenvalues/eigenvectors are computed using LAPACK/MAGMA routines ``_syevd``, ``_heevd``.
```
\`\`_syevd\`\` and \`\`_heevd\`\`

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+
+.. note:: The eigenvalues/eigenvectors are computed using LAPACK/MAGMA routines ``_syevd``, ``_heevd``.
+          This function always checks whether the call to LAPACK/MAGMA is successful
+          using ``info`` argument of ``_syevd``, ``_heevd``.
```
"using \`\`info\`\` argument of \`\`_syevd\`\`, \`\`_heevd\`\`. -> "and throws a RuntimeError if it isn't."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+.. note:: The eigenvalues/eigenvectors are computed using LAPACK/MAGMA routines ``_syevd``, ``_heevd``.
+          This function always checks whether the call to LAPACK/MAGMA is successful
+          using ``info`` argument of ``_syevd``, ``_heevd``.
+          For CUDA this causes a cross-device memory synchronization.
```
"For" -> "On"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+Args:
+    input (Tensor): the Hermitian :math:`n \times n` matrix or the batch
+                    of such matrices of size :math:`(*, n, n)` where `*` is one or more batch dimensions.
+    UPLO ('L', 'U', optional): controls whether to use upper-triangular or lower-triangular part
```
"to use upper-triangular or lower-triangular" -> "to use the upper-triangular or the lower-triangular"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,6 +140,129 @@
+This function computes the eigenvalues of a complex Hermitian (or real symmetric) matrix :attr:`input`
+or batches of such matrices. The eigenvalues are returned in ascending order.
+
+Since :attr:`input` is assumed to be consisting of Hermitian matrices,
```
Since these docstrings are so similar the comments above should apply here.

mruberry reviewed (commented):
Hey @IvanYashchuk!

This looks really good. I made a lot of document suggestions but they should be easy to fix so we can merge this soon.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-eig

IvanYashchuk committed:
Use same error message for syevd as for symeig when failed to converge

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Complex batched matmul now works on cuda

IvanYashchuk committed:
, -> and

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -97,7 +97,7 @@ static inline void batchCheckErrors(std::vector<int64_t>& infos, const char* nam
       if (strstr(name, "svd")) {
         AT_ERROR(name, ": the updating process of SBDSDC did not converge (error: ", info, ")");
-      } else if (strstr(name, "symeig")) {
+      } else if (strstr(name, "symeig") || strstr(name, "syevd")) {
```
Let's use the same error message as for `symeig` if LAPACK/MAGMA's algorithm doesn't converge. I couldn't find what kind of input matrix would trigger this though.

IvanYashchuk commented:
@mruberry, thank you for the feedback! I've fixed documentation issues.

I've added a small change to the source code https://github.com/pytorch/pytorch/pull/45526#discussion_r525058120

mruberry reviewed (approved):
Cool, looks great, @IvanYashchuk!

IvanYashchuk committed:
Safe use of std::toupper with plain chars

the argument should first be converted to unsigned char

IvanYashchuk commented:
Hey @mruberry, let's see if tests pass now.
I've modified the code to use static casts, now it should be safe as suggested. Also added missing `#include <cctype>`.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-eig

IvanYashchuk commented:
Resolved the accumulated merge conflicts.
@mruberry, can we try importing this again?

IvanYashchuk commented:
Autograd tests got broken with the last branch update. I'll fix that.

IvanYashchuk committed:
Fixed backward; can't in-place multiply real and complex tensors
